File: go-solutions/22/22.go
package main

var res []string

func generateParenthesis(n int) []string {
	res = make([]string, 0)
	gen(n, n, "")
	return res
}

func gen(left, right int, str string) {
	if left == 0 && right == 0 {
		res = append(res, str)
		return
	}

	if left > 0 {
		gen(left-1, right, str+"(")
	}

	if left < right {
		gen(left, right-1, str+")")
	}
}

File: cpp/22.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        vector<string> res;
        helper(res, "", n, 0);
        return res;
    }
    
    void helper(vector<string> &v, string str, int n, int m){
        if (n==0 && m==0){
            v.push_back(str);
            return;
        }
        if (m>0) helper(v, str+')', n,m-1);
        if (n>0) helper(v, str+'(', n-1, m+1);
    }
};

